include/deal.II/base/table_indices.h

```diff
@@ -43,11 +45,71 @@ class TableIndices
 public:
 
   /**
-   * Default constructor. It sets all indices to zero.
+   * Default constructor. This constructor sets all indices to zero.
    */
   TableIndices();
 
   /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<1> and initializes the single
+   * index with @p index0.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from one.
+   */
+  explicit TableIndices (const unsigned int index0);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<2> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from two.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<3> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from three.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<4> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from four.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<5> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from five.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3,
+                const unsigned int index4);
```

Comment:
What about something like
```
#include <array>

template <typename...> struct all_integers;

template <typename T> struct all_integers<T>
{
  static constexpr bool value = std::is_same<T, int>::value;
};

template <typename T, typename ...Rest> struct all_integers<T, Rest...>
{
  static constexpr bool value = std::is_same<T, int >::value && all_integers<Rest...>::value;
};

template <int N>
struct TableIndices
{ 
  template <typename... T, typename = typename std::enable_if<all_integers<T...>::value>::type>
  TableIndices (T... args)
    : values{args...}
    {
      static_assert (sizeof...(T) == N,
                     "Wrong number of constructor arguments!");
    } 
    
private:

  std::array<int , N> values;
};
```
if you really want to check the types of the template arguments? This gives error messages like

> test_table.cc:34:32: error: no matching function for call to ‘TableIndices<4>::TableIndices(int, int, int, double)’
   TableIndices<4> test(1,2,3,4.);